crates/pallet-feeds/src/feed_processor.rs

```diff
@@ -15,17 +15,82 @@
 
 //! Defines FeedProcessor and its types
 
+use crate::CallObject;
+use codec::{Compact, Decode};
 use sp_runtime::{DispatchError, DispatchResult};
 use sp_std::{vec, vec::Vec};
-use subspace_core_primitives::{crypto, Sha256Hash};
+use subspace_core_primitives::Sha256Hash;
 
-/// Object mapping that points to an object in a block
+/// Holds the offset to some portion of data within/or the object
 #[derive(Debug)]
-pub struct FeedObjectMapping {
-    /// Key scoped to the feed
-    pub key: Sha256Hash,
-    /// Offset of the data within object
-    pub offset: u32,
+pub enum FeedObjectMapping {
+    /// Maps the entire object treating it as content
+    /// If key is provided, it is namespaced to feed to avoid collisions
+    Object { key: Option<Vec<u8>> },
+    /// Maps some data within the object at the specific offset.
+    /// Data must be a scale encoded `Vec<u8>`
+    /// If key is provided, it namespaced to feed to avoid collisions
+    Content { key: Option<Vec<u8>>, offset: u32 },
+}
```

Comment:
>  We basically have 2 options which are the same with offset implied to be 0 in the first case, which you can already do with the second

No, there is a difference. First one hashes the whole object passed to the put with offset to 0. The offset directly points to the scale encoded object.

Second one maps to a specific content that is scale encoded within the object. So we always need to skip the bytes of the C<u32> the length of the object represent.

From the feed processor perpective, their base_offset always start after these length bytes.

Reply:
I am confused why there is a distinction between hashing the whole or part of the object. It should be exactly the same with the difference being `offset`.

Also we don't point to just SCALE-encoded content, we always point to the vector (in case of content-addressable data) and for custom we don't care at all what it is (strictly speaking), but generally expect it to be a vector too.

Instead of having 2 branches/options in suggested enum we have 4 and it is harder to explain what they do and why we need them.

Reply:
>  it is harder to explain what they do and why we need them.

Agreed!